fmp_server.py

```diff
@@ -38,7 +38,22 @@ def recvAnsConnection(self, packet_id, data, seq):
         connection_data = pati.ConnectionData.unpack(data)
         self.server.debug("Connection: {!r}".format(connection_data))
         self.sendNtcLogin(3, connection_data, seq)
-
+    
+    def recvReqShut(self, packet_id, data, seq):
```

Comment:
This is shadowing its previous implementation and should be implemented in the `pat` module instead.

Reply:
This is something that does need to be handled separately in `fmp_server`, since this is specifically an `fmp_server` `ReqShut` handler. `ReqShut` gets sent to each of the different servers at different times and for different reasons, so it still needs its own handler in the `pat` module, but this is (effectively) an `override` function for how we need to handle the `fmp_server` receiving it.

Reply:
I strongly believe that each shared packet has the same behaviour, if not there is a high chance that we're doing something wrong. That's not always true (like `recvAnsConnection`) but it often was.

What would not work if it's implemented in `pat` instead? What would be the missing data or the required one to make that works?

Reply:
I originally tried to implement it in `pat,` and realized that the other servers (the ones that are not `Fmp`) receive this exact same packet multiple times during the login process. While these do need to be responded to, thus the need for `pat` to have a simpler version of this implementation, the client sending this packet to `Fmp` specifically indicates a logging out of one of the game servers. This is the logout indicator that I was looking for that allowed me to remove the player from the game servers during a logout from the server select screen.

When I implemented it in `pat`, the result was the server experiencing an exception due to it removing players from the servers prematurely (during the login process.) One of the checks that I would have needed to do would have been to see that the server receiving it is `Fmp`, which is accomplished by using an override function in `fmp_server`.